File: tools/speak.py
#!/usr/bin/env python3
"""
Text-to-Speech tool using Windows Speech Synthesis - optimized for speed.
"""

import json
import subprocess
import sys


def speak_windows(text: str) -> bool:
    ps_script = f'Add-Type -AssemblyName System.Speech; $synth = New-Object System.Speech.Synthesis.SpeechSynthesizer; $synth.Speak("{text.replace('"', "'").replace(chr(92), chr(92)+chr(92))}")'
    try:
        subprocess.Popen(
            ['powershell', '-Command', ps_script],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL
        )
        return True
    except Exception:
        return False


def main() -> int:
    try:
        input_data = sys.stdin.read().strip()

        if not input_data:
            print(json.dumps({"success": False, "message": "No input provided."}))
            return 1

        data = json.loads(input_data)
        text = data.get("text", "").strip()

        if not text:
            print(json.dumps({"success": False, "message": "No text provided."}))
            return 1

        if not speak_windows(text):
            print(json.dumps({"success": False, "message": "Failed to speak."}))
            return 1

        print(json.dumps({"success": True, "message": "Speech played successfully."}))
        return 0

    except Exception as e:
        print(json.dumps({"success": False, "message": f"Error: {str(e)}"}))
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/speak.js
import { tool } from "@opencode-ai/plugin";
import { z } from "zod";
import { spawn } from "child_process";
import path from "path";
import os from "os";

const speakTool = tool({
  description: "Convert text to speech using Microsoft Edge TTS",
  args: {
    text: z.string().describe("The text to convert to speech"),
    voice: z
      .string()
      .default("en-US-AriaNeural")
      .describe("The voice to use for speech synthesis"),
    rate: z
      .string()
      .default("+0%")
      .describe("The speech rate adjustment"),
    volume: z
      .string()
      .default("+0%")
      .describe("The speech volume adjustment"),
  },
  async execute(args) {
    const { text, voice, rate, volume } = args;

    const homeDir = os.homedir();
    const pythonScriptPath = path.join(homeDir, ".config", "opencode", "tools", "speak.py");

    const inputJson = JSON.stringify({
      text,
      voice,
      rate,
      volume,
    });

    const result = await new Promise((resolve, reject) => {
      const pythonProcess = spawn("python", [pythonScriptPath], {
        stdio: ["pipe", "pipe", "pipe"],
      });

      let stdout = "";
      let stderr = "";

      pythonProcess.stdout.on("data", (data) => {
        stdout += data.toString();
      });

      pythonProcess.stderr.on("data", (data) => {
        stderr += data.toString();
      });

      pythonProcess.on("close", (code) => {
        if (code === 0) {
          try {
            const result = JSON.parse(stdout.trim());
            resolve(result);
          } catch {
            resolve({
              success: false,
              message: `Failed to parse output: ${stdout}`,
            });
          }
        } else {
          resolve({
            success: false,
            message: stderr || `Script exited with code ${code}`,
          });
        }
      });

      pythonProcess.on("error", (error) => {
        resolve({
          success: false,
          message: `Failed to execute: ${error.message}`,
        });
      });

      pythonProcess.stdin.write(inputJson);
      pythonProcess.stdin.end();
    });

    if (!result.success) {
      throw new Error(result.message);
    }

    return result.message;
  },
});

export default speakTool;
